File: app/src/main/java/co/smartreceipts/android/model/impl/columns/AbstractColumnImpl.java
package co.smartreceipts.android.model.impl.columns;

import android.support.annotation.NonNull;

import java.util.List;

import co.smartreceipts.android.model.Column;
import co.smartreceipts.android.sync.model.SyncState;

/**
 * Provides an abstract implementation of the column contract to cover shared code
 */
public abstract class AbstractColumnImpl<T> implements Column<T> {

    private final int mId;
    private final String mName;
    private final SyncState mSyncState;
    private final int mCustomOrderId;

    public AbstractColumnImpl(int id, @NonNull String name, @NonNull SyncState syncState) {
        this(id, name, syncState, 0);
    }

    public AbstractColumnImpl(int id, @NonNull String name, @NonNull SyncState syncState, int customOrderId) {
        mId = id;
        mName = name;
        mSyncState = syncState;
        mCustomOrderId = customOrderId;
    }

    @Override
    public int getId() {
        return mId;
    }

    @Override
    public int getCustomOrderId() {
        return mCustomOrderId;
    }

    @Override
    @NonNull
    public final String getName() {
        return mName;
    }

    @Override
    @NonNull
    public String getHeader() {
        return getName();
    }

    @Override
    @NonNull
    public String getFooter(@NonNull List<T> rows) {
        return "";
    }

    @Override
    public int compareTo(@NonNull Column otherColumn) {
        return getId() - otherColumn.getId();
    }

    @NonNull
    @Override
    public SyncState getSyncState() {
        return mSyncState;
    }

    @Override
    public String toString() {
        return getName();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AbstractColumnImpl)) return false;

        AbstractColumnImpl that = (AbstractColumnImpl) o;

        if (mId != that.mId) return false;
        if (mCustomOrderId != that.mCustomOrderId) return false;
        if (!mName.equals(that.mName)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = mId;
        result = 31 * result + mCustomOrderId;
        result = 31 * result + mName.hashCode();
        return result;
    }
}

File: app/src/main/java/co/smartreceipts/android/model/impl/columns/AbstractColumnImpl.kt
package co.smartreceipts.android.model.impl.columns

import co.smartreceipts.android.model.ActualColumnDefinition
import co.smartreceipts.android.model.Column
import co.smartreceipts.android.model.Keyed
import co.smartreceipts.core.sync.model.SyncState
import java.util.*

/**
 * Provides an abstract implementation of the column contract to cover shared code
 */
abstract class AbstractColumnImpl<T> @JvmOverloads constructor(
    override val id: Int,
    private val columnDefinition: ActualColumnDefinition,
    override val syncState: SyncState,
    override val customOrderId: Long = 0,
    override val uuid: UUID = Keyed.MISSING_UUID
) : Column<T> {

    override val type: Int = columnDefinition.columnType

    override val headerStringResId = columnDefinition.columnHeaderId

    override fun getFooter(rows: List<T>) = ""

    override operator fun compareTo(other: Column<T>): Int =
        if (customOrderId == other.customOrderId) {
            id - other.id
        }
        else java.lang.Long.compare(customOrderId, other.customOrderId)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as AbstractColumnImpl<*>?
        return id == that!!.id &&
                uuid == that.uuid &&
                customOrderId == that.customOrderId &&
                columnDefinition == that.columnDefinition
    }

    override fun hashCode(): Int = Objects.hash(id, uuid, columnDefinition, customOrderId)
}
